src/core/ext/transport/chttp2/transport/ping_promise.cc

```diff
@@ -0,0 +1,150 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+#include "src/core/ext/transport/chttp2/transport/ping_promise.h"
+
+#include "src/core/lib/promise/if.h"
+#include "src/core/lib/promise/latch.h"
+#include "src/core/lib/promise/map.h"
+#include "src/core/lib/promise/race.h"
+#include "src/core/lib/promise/sleep.h"
+#include "src/core/lib/promise/try_seq.h"
+#include "src/core/util/match.h"
+#include "src/core/util/time.h"
+
+namespace grpc_core {
+namespace http2 {
+using SendPingArgs = ::grpc_core::http2::PingSystemInterface::SendPingArgs;
+using Callback = absl::AnyInvocable<void()>;
+
+#define PING_LOG                                           \
+  LOG_IF(INFO, (GRPC_TRACE_FLAG_ENABLED(http) ||           \
+                GRPC_TRACE_FLAG_ENABLED(bdp_estimator) ||  \
+                GRPC_TRACE_FLAG_ENABLED(http_keepalive) || \
+                GRPC_TRACE_FLAG_ENABLED(http2_ping)))
+
+Promise<absl::Status> PingSystem::PingPromiseCallbacks::RequestPing(
+    Callback on_initiate) {
+  std::shared_ptr<Latch<void>> latch = std::make_shared<Latch<void>>();
+  auto on_ack = [latch]() { latch->Set(); };
+  ping_callbacks_.OnPing(std::move(on_initiate), std::move(on_ack));
+  return Map(latch->Wait(), [latch](Empty) { return absl::OkStatus(); });
+}
+
+Promise<absl::Status> PingSystem::PingPromiseCallbacks::WaitForPingAck() {
+  std::shared_ptr<Latch<void>> latch = std::make_shared<Latch<void>>();
+  auto on_ack = [latch]() { latch->Set(); };
+  ping_callbacks_.OnPingAck(std::move(on_ack));
+  return Map(latch->Wait(), [latch](Empty) { return absl::OkStatus(); });
+}
+
+// Ping System implementation
+PingSystem::PingSystem(
+    const ChannelArgs& channel_args,
+    std::unique_ptr<PingSystemInterface> ping_interface,
+    std::shared_ptr<grpc_event_engine::experimental::EventEngine> event_engine)
+    : ping_callbacks_(event_engine),
+      ping_abuse_policy_(channel_args),
+      ping_rate_policy_(channel_args, /*is_client=*/true),
+      ping_interface_(std::move(ping_interface)) {}
+
+void PingSystem::TriggerDelayedPing(Duration wait, Party* party) {
+  // Spawn at most once.
+  if (delayed_ping_spawned_) {
+    return;
+  }
+  delayed_ping_spawned_ = true;
+  party->Spawn(
+      "DelayedPing",
+      [this, wait]() mutable {
+        VLOG(2) << "Scheduling delayed ping after wait=" << wait;
+        return AssertResultType<absl::Status>(TrySeq(
+            Sleep(wait),
+            [this]() mutable { return ping_interface_->TriggerWrite(); }));
+      },
+      [this](auto) { delayed_ping_spawned_ = false; });
+}
+
+bool PingSystem::NeedToPing(Duration next_allowed_ping_interval, Party* party) {
+  bool send_ping_now = false;
```

Comment:
Suggest dropping this variable... line 85 can be `return false`, line 88 can be `return Match(...)`